### PR TITLE
[GraphQL/Data] Refactor PgManager and EXPLAIN-based costing

### PR DESCRIPTION
**Pull request by amnn**
## Description

Starting to refactor the data layer to use the cursor framework. This refactor will extend beyond directly plugging in the framework, so that it's possible to consolidate the logic related to handling pagination into a single place, rather than repeating it for every query.

Refactorings include:

Moving the logic related to the database into its own module, rather than `context_data` (which was intended just for things that are added to the context).

Flipping the architecture on its head. Currently, the architecture is based on:

- A single query builder type that contains boxed queries for every relevant type.

- A single interface to implement for each DB backend that deals with:
  - Calculating the estimated EXPLAIN cost and logging it.
  - Running queries through the `IndexerReader`.
  - Pagination (implemented for each endpoint that requires pagination).
  - Converting Indexer/DB types into GraphQL types.

This design has a couple of drawbacks:

- The query builders and implementations ended up being "God classes", and grow to include many dependencies, also drawing logic for how to fetch a type, away from that type.

- Pagination logic is repeated for every query/type that needs it, making it hard to replace that logic in a consistent way (there are also some existing issues with how these endpoints implement previous and next page detection which are hard to fix uniformly for this reason).

- The database backend being used is not fully encapsulated -- it is referenced in both the PG-specific implementation of the query builder, and the outer "data provider" interface.

The proposed architecture is as follows:

- A `QueryExecutor` interface that encapsulates the DB backend being used, and a new implementation of `PgManager` that implements it for Postgres.

- This interface exposes a generic way to run queries that read one or many results from the DB (Similar to `diesel`'s `RunQueryDsl`).

- Queries are passed in as thunks that return a generic boxed query type (as in the current design), but this type is generic on query source (unlike the current query builder interface which specifies which table each boxed query is from).

- EXPLAIN-based costing (or now, logging) has been made an implementation detail of the postgres-specific implementation, rather than being exposed as a generic facility (because the logic for how to interpret EXPLAIN output is DB-backend-specific). This will make it easier to implement a sqlite backend that does not care about `EXPLAIN` (for use when the GraphQL service and indexer are embedded in a single process).

- Logic for how to fetch each GraphQL type will move to that type's module, to split apart the existing "God module".

- Logic for pagination will move into the `Cursor`/`Page` types.

This PR introduces the new home for this logic (the `data` module), the `QueryExecutor` trait, an implementation for the Pg Backend that implements `EXPLAIN`-based logging and a test for the same.  The rest of the architecture and integration will follow (along with use of the cursor framework and renames of Connection-based queries).

It does not delete any of the existing logic, or do any integration, so the new implementation is called `PgManager_` (note the trailing underscore). Which will duplicate the existing `PgManager` until it can take over all its responsibilities (at which point, it can be deleted).

## Test Plan

New integration test for EXPLAIN-based cost estimation:

```
sui-graphql-rpc$ cargo nextest run \
  -j 1 --features pg_integration   \
  -- data::pg::test_query_cost
```

## Stack
- #15467 
- #15470 
- #15471 
- #15472 
- #15473
- #15474 
- #15475 
- #15484 
- #15485
- #15519
- #15520
- #15521
- #15522
- #15523

### CODE REVIEW

**Review by wlmyng — APPROVED** [review]
I like this a lot!

> Queries are passed in as thunks that return a generic boxed query type (as in the current design), but this type is generic on query source (unlike the current query builder interface which specifies which table each boxed query is from).

Incredible, I had wanted to do this, but couldn't wrap my head around how to type things as a BoxedQuery agnostic of the db backend. Really nice

> Logic for how to fetch each GraphQL type will move to that type's module, to split apart the existing "God module".

What started out as a compilation of query snippets to fetch the various Stored types ended up becoming god :painlaugh: I think returning the responsibility back to the type modules make sense, and I suppose we can always lift repeated dsl snippets later

**Review by stefan-mysten — COMMENTED** [review]
Uhm, that's a lot of type magic :D Lovely work @amnn.

File: crates/sui-graphql-rpc/Cargo.toml [hunk 1]
```diff
@@ -72,6 +72,7 @@ expect-test.workspace = true
 hyper.workspace = true
 insta.workspace = true
 serde_json.workspace = true
+sui-framework.workspace = true
```
Comment (stefan-mysten) [inline crates/sui-graphql-rpc/Cargo.toml:75]: Why is this needed?

Comment (amnn) [reply crates/sui-graphql-rpc/Cargo.toml:75]: I added a test for query costing that needed some realistic objects to load into the DB and then fetch back out, so I decided to use the system packages!  (Note, this is a dev-dependency).

**Comment by amnn** [comment]
Thanks for the fast reviews both!

> Uhm, that's a lot of type magic :D

It is indeed, but not necessarily something to aspire to 😅 this is mostly a function of how Diesel is set-up, and I was just bashing away at it incrementally until the compiler stopped shouting at me.

File: crates/sui-graphql-rpc/Cargo.toml (hunk 1 above)
Comment (stefan-mysten) [reply crates/sui-graphql-rpc/Cargo.toml:75]: Oh, I missed the part of it being a `dev-dependency`.

File: crates/sui-graphql-rpc/src/data.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    where
+        Q: Fn() -> BoxedQuery<ST, QS, Self, GB>,
+        BoxedQuery<ST, QS, Self, GB>: LoadQuery<'static, Self::Connection, U>,
+        BoxedQuery<ST, QS, Self, GB>: QueryFragment<<Self::Connection as Connection>::Backend>,
+        QS: QuerySource,
+        Q: Send + 'static,
+        U: Send + 'static;
+
+    /// Helper to limit a query that fetches multiple values to return only its first value. `query`
+    /// is at hink that returns a boxed query when called.
```
[the remark this answers is not in the document]
Comment (amnn) [reply crates/sui-graphql-rpc/src/data.rs:74]: Resolved during staging for land!
